File: ruby/elements/expect.rb
require_relative '../utils'

module Elements
  class Expect < Element
    include Utils

    @@number_of_tests = 0

    def initialize( compiler, article, lines)
      @@number_of_tests += 1
      if lines.size != 2
        article.error( 'Bad expect directive')
        @title, @regex = '', /./
      else
        @title = lines[0]
        expr, @negate = lines[1], false
        if m = /^\!(.*)$/.match( expr)
          expr, @negate = m[1], true
        end
        @regex = Regexp.new( expr.strip, Regexp::MULTILINE)
      end
    end

    def self.number_of_tests
      @@number_of_tests
    end

    def page_content?
      false
    end

    def post_process_html( article, html)
      if @regex =~ html
        article.error( @title) if @negate
      else
        article.error( @title) unless @negate
      end
      html
    end
  end
end

File: java/src/com/alofmethbin/jadawin/elements/Expect.java
package com.alofmethbin.jadawin.elements;

import com.alofmethbin.jadawin.Article;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Expect extends Element {
    private static int numberOfTests = 0;
    private boolean negate = false;
    private String title, match;
    
    public Expect( Article article, List<String> lines) {
        super( article);
        synchronized ( Expect.class ) {
            numberOfTests ++;
        }

        if (lines.size() != 2) {
            error( "Bad error directive");
        } else {
            title = lines.get(0);
            match = lines.get(1);
        }
    }

    @Override
    public boolean isContent() {
        return false;
    }

    public static int numberOfTests() {
        return numberOfTests;
    }

    @Override
    public String postProcessHTML( String html) {
        Matcher m = Pattern.compile( match, Pattern.MULTILINE).matcher( html);
        if (m.find() ^ negate) {
            error( title);
        }
        return html;
    }
}
